test/subtype.jl

```diff
@@ -1786,3 +1786,10 @@ let X1 = Tuple{AlmostLU, Vector{T}} where T,
     @test I >: actual
     @test_broken I == actual
 end
+
+# issue #22787
+# for now check that these don't stack overflow
+@test_broken typeintersect(Tuple{Type{Q}, Q, Ref{Q}} where Q<:Ref,
+                           Tuple{Type{S}, Union{Ref{S}, Ref{R}}, R} where R where S) != Union{}
+@test_broken typeintersect(Tuple{Type{T}, T, Ref{T}} where T,
+                           Tuple{Type{S}, Ref{S}, S} where S) != Union{}
```

Comment:
If I'm not mistaken, `@test_broken` passes, even if they do stack overflow:

Here's an example of these tests passing on master, despite the stack overflow:
```julia
julia> # for now check that these don't stack overflow
       @test_broken typeintersect(Tuple{Type{Q}, Q, Ref{Q}} where Q<:Ref,
                                  Tuple{Type{S}, Union{Ref{S}, Ref{R}}, R} where R where S) != Union{}
Test Broken
  Expression: typeintersect(Tuple{Type{Q}, Q, Ref{Q}} where Q <: Ref, (Tuple{Type{S}, Union{Ref{S}, Ref{R}}, R} where R) where S) != Union{}

julia> @test_broken typeintersect(Tuple{Type{T}, T, Ref{T}} where T,
                                  Tuple{Type{S}, Ref{S}, S} where S) != Union{}
Test Broken
  Expression: typeintersect(Tuple{Type{T}, T, Ref{T}} where T, Tuple{Type{S}, Ref{S}, S} where S) != Union{}

julia> # You can see that it's stack overflowing when we remove the `@test_broken`:
       @test typeintersect(Tuple{Type{Q}, Q, Ref{Q}} where Q<:Ref,
                                  Tuple{Type{S}, Union{Ref{S}, Ref{R}}, R} where R where S) != Union{}
Error During Test at REPL[4]:2
  Test threw exception
  Expression: typeintersect(Tuple{Type{Q}, Q, Ref{Q}} where Q <: Ref, (Tuple{Type{S}, Union{Ref{S}, Ref{R}}, R} where R) where S) != Union{}
  StackOverflowError:
  Stacktrace:
    [1] typeintersect(a::Any, b::Any)
      @ Base ./reflection.jl:563
    [2] top-level scope
      @ REPL[4]:2
    [3] eval(m::Module, e::Any)
      @ Core ./boot.jl:360
...
```

Reply:
Maybe you can change to something like this, which _would_ fail if they stack overflow, but still allow the broken intersection test? Thanks!

```suggestion
@testset begin
    # These tests don't pass, but for now check that at least they don't stack overflow
    t = typeintersect(Tuple{Type{Q}, Q, Ref{Q}} where Q<:Ref,
                           Tuple{Type{S}, Union{Ref{S}, Ref{R}}, R} where R where S)
    @test_broken t != Union{}
    t = typeintersect(Tuple{Type{T}, T, Ref{T}} where T,
                           Tuple{Type{S}, Ref{S}, S} where S)
    @test_broken t != Union{}
end
```

Reply:
The above suggestion passes for me on this branch, and fails on master.

Reply:
To stop my head spinning... what's the correct answer to these intersection problems?

Reply:
Here's some:
```
julia> Tuple{Type{Ref}, Ref{Ref}, Ref{Ref}} <: Tuple{Type{T}, Ref{T}, T} where T
true
julia> Tuple{Type{Ref}, Ref{Ref}, Ref{Ref}} <: Tuple{Type{T}, T, Ref{T}} where T
true

julia> Tuple{Type{Ref{<:Ref}}, Ref{Ref{<:Ref}}, Ref{Ref{<:Ref}}} <: Tuple{Type{T}, T, Ref{T}} where T
true
julia> Tuple{Type{Ref{<:Ref}}, Ref{Ref{<:Ref}}, Ref{Ref{<:Ref}}} <: Tuple{Type{T}, T, Ref{T}} where T
true

julia> Tuple{Type{Ref{<:Ref{<:Ref}}}, Ref{Ref{<:Ref{<:Ref}}}, Ref{Ref{<:Ref{<:Ref}}}} <: Tuple{Type{T}, T, Ref{T}} where T
true
julia> Tuple{Type{Ref{<:Ref{<:Ref}}}, Ref{Ref{<:Ref{<:Ref}}}, Ref{Ref{<:Ref{<:Ref}}}} <: Tuple{Type{T}, Ref{T}, T} where T
true
```
I don't know if there's a general closed-form expression of this, since it seems to be the Union of this out to infinity?